src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+            nailgun_args.push(arg.clone());
+        } else if have_seen_classpath && !have_processed_classpath {
+            // Process the classpath string
+            nailgun_args.push(arg.clone());
+            have_processed_classpath = true;
+        } else if have_processed_classpath && !arg.starts_with("-") {
+            // Process the main class:
+            // I have already seen the value of the -cp classpath, and this is not a flag.
+            client_args.push(arg.clone());
+            have_seen_main_class = true;
+        } else {
+            // Process the rest (jvm options to pass to nailgun)
+            nailgun_args.push(arg.clone());
+        }
+    }
+    ParsedArgLists {
+        nailgun_args: nailgun_args,
+        client_args: client_args
+    }
+}
+
+// TODO(8481) The input_files arg should be calculated from deeply fingerprinting the classpath.
+fn get_nailgun_request(args: Vec<String>, _input_files: Digest, jdk: Option<PathBuf>) -> ExecuteProcessRequest {
+    let mut full_args = args;
+    full_args.push(NAILGUN_MAIN_CLASS.to_string());
+    full_args.extend(ARGS_TO_START_NAILGUN.iter().map(|a| a.to_string()));
+
+    ExecuteProcessRequest {
+        argv: full_args,
+        env: BTreeMap::new(),
+        input_files: hashing::EMPTY_DIGEST,
+        output_files: BTreeSet::new(),
+        output_directories: BTreeSet::new(),
+        timeout: Duration::new(1000, 0),
+        description: String::from("ExecuteProcessRequest to start a nailgun"),
+        unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule: hashing::EMPTY_DIGEST,
+        jdk_home: jdk,
+        target_platform: Platform::Darwin,
+        is_nailgunnable: true,
+    }
+}
+
+///
+/// A command runner that can run local requests under nailgun.
+///
+/// It should only be invoked with local requests.
+/// It will read a flag marking an `ExecuteProcessRequest` as nailgunnable.
+/// If that flag is set, it will connect to a running nailgun server and run the command there.
+/// Otherwise, it will just delegate to the regular local runner.
+///
+pub struct NailgunCommandRunner {
+    inner: Arc<super::local::CommandRunner>,
+    nailguns: NailgunPool,
+    metadata: ExecuteProcessRequestMetadata,
+    workdir_base: PathBuf,
+    python_distribution_absolute_path: PathBuf,
+}
+
+impl NailgunCommandRunner {
+    pub fn new(runner: super::local::CommandRunner, metadata: ExecuteProcessRequestMetadata, python_distribution_absolute_path: PathBuf) -> Self {
+        NailgunCommandRunner {
+            inner: Arc::new(runner),
+            nailguns: NailgunPool::new(),
+            metadata: metadata,
+            workdir_base: std::env::temp_dir(),
+            python_distribution_absolute_path: python_distribution_absolute_path,
+        }
+    }
+
+    // Ensure that the workdir for the given nailgun name exists.
+    fn get_nailguns_workdir(&self, nailgun_name: &NailgunProcessName) -> Result<PathBuf, String> {
```

Comment:
Half of this function appears to be guaranteeing `workdir` exists, and half that `workdir_base` exists?

Reply:
Added tests, and fixed it.